test/integration/fake_upstream.cc

```diff
@@ -585,14 +608,27 @@ FakeUpstream::waitForHttpConnection(Event::Dispatcher& client_dispatcher,
                 client_dispatcher, 5ms)) {
           continue;
         }
-        connection = std::make_unique<FakeHttpConnection>(
-            upstream, upstream.consumeConnection(), upstream.http_type_, upstream.timeSystem(),
-            Http::DEFAULT_MAX_REQUEST_HEADERS_KB, Http::DEFAULT_MAX_HEADERS_COUNT,
-            envoy::config::core::v3::HttpProtocolOptions::ALLOW);
       }
-      VERIFY_ASSERTION(connection->initialize());
-      VERIFY_ASSERTION(connection->readDisable(false));
-      return AssertionSuccess();
+
+      AssertionResult result = AssertionSuccess();
+      absl::Notification done;
+      ASSERT(!upstream.dispatcher_->isThreadSafe());
+      upstream.dispatcher_->post([&]() {
+        {
+          absl::MutexLock lock(&upstream.lock_);
+          connection = std::make_unique<FakeHttpConnection>(
+              upstream, upstream.consumeConnection(), upstream.http_type_, upstream.timeSystem(),
+              Http::DEFAULT_MAX_REQUEST_HEADERS_KB, Http::DEFAULT_MAX_HEADERS_COUNT,
+              envoy::config::core::v3::HttpProtocolOptions::ALLOW);
+          result = connection->initialize();
```

Comment:
You don't like the VERIFY_ASSERTION macro in use before? I actually didn't even know about it but it seems in the style of c++ gtest.

Reply:
It is not about liking, it is about this lambda not being able to return an AssertResult.

Reply:
Ah I see. Well I think the abstraction I suggested above would help; you'd still need to use a `result` var but it would be within the implementation of dispatchAndWait, which could then return it.